# Lendo os dados e gravando em múltiplos formatos

Arquivo de dados:

https://drive.google.com/file/d/1U0e-ssOerAZwEL9-6wBduj

In [23]:
import csv, io, json, requests, sqlite3
import pandas as pd

# O arquivo foi baixado para ../data/bkp/AppleStore.csv
df = pd.read_csv('../data/bkp/AppleStore.csv', index_col=0)

# Execução 1: tag "news" com maior quantidade de avaliações

Identifique a Aplicação da categoria News, que tiver a maior quantidade de avaliações rating_count_tot.

In [32]:
df_news = df[df['prime_genre'] == 'News'].filter(['track_name', 'rating_count_tot'])

# Verificando se pode haver mais de 1 ocorrência de track_name (versões diferentes)
df_news.groupby('track_name').count().sort_values(by='rating_count_tot', ascending=False)

rating_count_tot
track_name                                                          
20 Minutes.fr - l'actualité en continu                             1
franceinfo - l'actualité & les élections en direct                 1
WSVN Hurricane Tracker                                             1
WIRED Magazine                                                     1
ViRATES[バイレーツ]-面白ネタまとめの決定版！                                        1
Verbrechen - echte Polizeifälle aus Deiner Umge...                 1
USA TODAY                                                          1
US Presidential Election 2016 - Polls                              1
Twitter                                                            1
TopBuzz: Best Viral Videos, GIFs, TV & News                        1
Ticket Scanner for Powerball & MegaMillions Pool                   1
The Washington Post Classic                                        1
The Guardian                                                       1
Tagesschau                                                         1
State Council - Official Chinese government app                    1
SmartNews - Trending News & Stories                                1
Smart Channel -New Style of News Reader-                           1
ZAKER 专业版                                                          1
n-tv Nachrichten                                                   1
Scanner911 Pro                                                     1
theSkimm                                                           1
腾讯新闻HD-最资深的阅读软件                                                    1
腾讯新闻-头条新闻热点资讯掌上阅读软件                                                1
网易新闻 - 精选好内容，算出你的兴趣                                                1
新浪新闻-阅读最新时事热门头条资讯视频                                                1
搜狐新闻—新闻热点资讯掌上阅读软件                                                  1
天天快报 - 最热门的新闻资讯软件                                                  1
华尔街见闻（专业版）-全球财经新闻精选                                                1
凤凰新闻(专业版)-有料的军事新闻、娱乐短视频                                            1
今日头条(专业版) - 推荐热点新闻资讯、娱乐视频                                          1
...                                                              ...
Daily Classifieds for iPhone                                       1
Conservative Talk Radio                                            1
Clown Spotter - Find Clowns Around You                             1
CNN: Breaking US & World News, Live Video                          1
CNN Politics                                                       1
CBS News - Watch Free Live Breaking News                           1
Boycott Trump Biz                                                  1
BFMTV : l'info en continu                                          1
AOL: News, Email, Weather & Video                                  1
ABC News - US & World News + Live Video                            1
Fresco — Be a part of the news                                     1
HuffPost - News, Politics & Entertainment                          1
ST channel ［エスティーチャンネル］- 雑誌『セブンティーン』公式アプリ                          1
JCnews - Anime & Game Culture                                      1
SPIEGEL ONLINE - Nachrichten                                       1
Reddit Official App: All That's Trending and Viral                 1
Quartz • News in a whole new way                                   1
Presidential Election & Electoral College Maps                     1
PodCruncher podcast app - Player and manager fo...                 1
Pocket Casts                                                       1
PS Deals+ - Games Price Alerts for PS4, PS3, Vita                  1
OPM Alert                                                          1
News Pro - Breitbart Edition                                       1
News Break - Local & World Breaking News & Radio                   1
NBC News                                                           1
MSNBC                                             

In [42]:
# Obtendo a aplicação com maior número de ratings em sua história
df_news.sort_values(by='rating_count_tot', ascending=False).head(1)

track_name  rating_count_tot
251    Twitter            354058

## Resultado: Twitter, com 354058 avaliações

# Execução 2: 10 mais avaliados entre as tags "music" e "book"

Identificar quais são as 10 Aplicações do gênero Music e Book que possuem a maior quantidade de avaliações no arquivo csv apple_store.

In [46]:
df_music_book = df[df['prime_genre'].isin(['Music', 'Book'])]

df_topten = df_music_book.sort_values(by='rating_count_tot', ascending=False).head(10)
df_topten

id                                         track_name  \
8     284035177                            Pandora - Music & Radio   
202   324684580                                      Spotify Music   
20    284993459  Shazam - Discover music, artists, videos & lyrics   
40    290638154          iHeartRadio – Free Music & Radio Stations   
93    302584613        Kindle – Read eBooks, Magazines & Textbooks   
270   336353151                         SoundCloud - Music & Audio   
816   421254504                               Magic Piano by Smule   
1431  509993510                                        Smule Sing!   
803   418987775   TuneIn Radio - MLB NBA Audiobooks Podcasts Music   
1438  510855668                                       Amazon Music   

      size_bytes currency  price  rating_count_tot  rating_count_ver  \
8      130242560      USD    0.0           1126879              3594   
202    132510720      USD    0.0            878563              8253   
20     147093504      USD    0.0            402925               136   
40     116443136      USD    0.0            293228               110   
93     169747456      USD    0.0            252076                80   
270    105009152      USD    0.0            135744               594   
816     55030784      USD    0.0            131695              1102   
1431   109940736      USD    0.0            119316                33   
803    101735424      USD    0.0            110420               370   
1438    77778944      USD    0.0            106235              4605   

      user_rating  user_rating_ver     ver cont_rating prime_genre  \
8             4.0              4.5   8.4.1         12+       Music   
202           4.5              4.5   8.4.3         12+       Music   
20            4.0              4.5  11.0.3         12+       Music   
40            4.0              3.0   8.0.0         12+       Music   
93            3.5              4.5    5.11          4+        Book   
270           4.0              4.5   5.6.0          4+       Music   
816           4.5              4.0   8.3.1          4+       Music   
1431          4.5              4.0   5.2.3          9+       Music   
803           4.5              4.5    11.9         12+       Music   
1438          4.5              5.0   6.5.0          4+       Music   

      sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  
8                  37                4         1        1  
202                37                5        18        1  
20                 37                3        16        1  
40                 37                5         2        1  
93                 37                5         9        1  
270                37                4         7        1  
816                37                5        12        1  
1431               37                5        12        1  
803                37                4        14        1  
1438               37                5         6        1

# Execução 3: 10 mais citados entre as tags "music" e "book"

Após encontrar a aplicação do tipo News utilize a sua API, para identificar quais das 10 aplicações do tipo Music e Book, possuem o maior número de citações nessa API.

Consultar o número de citações de cada track do item anterior na API aplicação mais avaliada (twitter).

Não há indicativo de usuário no dataframe original. Uma alternativa seria buscar por parte do nome em # ou @ em um conjunto de tweets. No entanto, isso provocaria um efeito colateral, com aumento no número real em casos como Shazam (que adicionaria as menções e citações ao herói).

Em função disso, busquei manualmente os usuários no contexto dos aplicativos e então fiz uma busca pelas citações (@usuario), adicionando os resultados ao dataframe.

In [48]:
dict_users = {
    "id": [284035177, 324684580, 284993459, 290638154, 302584613, 336353151, 421254504, 509993510, 418987775, 510855668],
    "twitter_user": ["pandoramusic", "Spotify", "Shazam", "iHeartRadio", "AmazonKindle", "SoundCloud", "MagicPianoApp", "smule", "tunein", "amazonmusic"] 
}
df_mention = pd.DataFrame.from_dict(dict_users)

# Pesquisa pelos mentions na API do Twitter: n_citacoes

# Faz o merge com o top10
df_topten = df_topten.merge(df_mention)
df_topten

id   twitter_user
0  284035177   pandoramusic
1  324684580        Spotify
2  284993459         Shazam
3  290638154    iHeartRadio
4  302584613   AmazonKindle
5  336353151     SoundCloud
6  421254504  MagicPianoApp
7  509993510          smule
8  418987775         tunein
9  510855668    amazonmusic

In [27]:
df

id                                         track_name  \
1       281656475                                    PAC-MAN Premium   
2       281796108                          Evernote - stay organized   
3       281940292    WeatherBug - Local Weather, Radar, Maps, Alerts   
4       282614216  eBay: Best App to Buy, Sell, Save! Online Shop...   
5       282935706                                              Bible   
6       283619399                                   Shanghai Mahjong   
7       283646709             PayPal - Send and request money safely   
8       284035177                            Pandora - Music & Radio   
9       284666222                        PCalc - The Best Calculator   
10      284736660                                        Ms. PAC-MAN   
11      284791396                          Solitaire by MobilityWare   
12      284815117                                   SCRABBLE Premium   
13      284815942               Google – Search made just for mobile   
14      284847138                   Bank of America - Mobile Banking   
15      284862767                                           FreeCell   
16      284876795             TripAdvisor Hotels Flights Restaurants   
17      284882215                                           Facebook   
18      284910350     Yelp - Nearby Restaurants, Shopping & Services   
20      284993459  Shazam - Discover music, artists, videos & lyrics   
21      285005463                      Crash Bandicoot Nitro Kart 3D   
22      285946052                                             iQuran   
23      285994151                                        :) Sudoku +   
24      286058814    Yahoo Sports - Teams, Scores, News & Highlights   
25      286070473                          Mileage Log | Fahrtenbuch   
27      286799607                        Cleartune - Chromatic Tuner   
28      286906691          Lifesum – Inspiring healthy lifestyle app   
29      286911400                                           Hangman.   
31      288113403      iTranslate - Language Translator & Dictionary   
32      288120394                                           TouchOSC   
33      288419283                                         RadarScope   
...           ...                                                ...   
10995  1182265441                      脱出ゲーム　わたしをみつけて　-おじいさんとわたしの物語-   
10998  1182331762                      Escape from the frigid Igloo.   
11002  1182568288        Talking Santa - Video santa claus calls you   
11010  1183234072                                          CTFxCmoji   
11013  1183260922                    Room Escape Game - Santa's Room   
11016  1183548754                               Rescue the Enchanter   
11019  1183709176                                 My Diary - 你的名字非官方   
11022  1183856228  VR Thrills: Roller Coaster 360 (Google Cardboard)   
11024  1183986102   Santa Kids Hair Salon - Christmas Makeover Games   
11027  1184711626                                 Human Juggling Cup   
11031  1184800011                           Again - room escape game   
11033  1185209084                                  Saloons Unleashed   
11035  1185328193             Fam — Group video calling for iMessage   
11036  1185365336                   Laurie Hernandez the Human Emoji   
11038  1185428381                                               剑倚手游   
11040  1185538497                                  camera for filter   
11041  1185580782                                    Survivalcraft 2   
11042  1185731859                                    剑客情缘-高爆率高掉落天天疯玩   
11043  1185777521                                     问仙奇遇-新玩法新套装嗨到爆   
11050  1186108496                 脱出ゲーム - 書道教室 -  "漢字"の謎に満ちた部屋からの 脱出   
11051  1186126548                          Escape Game: illumination   
11060  1186384912       Demolition Derby Virtual Reality (VR) Racing   
11074  1187128255                         飞刀传奇-动作武侠热血江湖即时PK传奇（登录爆金装）   
11077  1187279979                       Add-Ons Studio 

# Entregável: Gravando em múltiplos formatos

O output esperado é a criação de um CSV, um JSON e uma base de dados local com as respectivas colunas: id, track_name, n_citacoes, size_bytes, price, prime_genre. Os dados relativos às Aplicações estão disponíveis no arquivo abaixo.


In [ ]:
# Filtro do dataframe para gravar em diversos formatos com um número limitado de colunas
df_table = df.filter(['id', 'track_name', 'n_citacoes', 'size_bytes', 'price', 'prime_genre'])
df_table.set_index('id')

# [ENTREGÁVEL] Criação de uma tabela no SQLite em memória
conn = sqlite3.connect(":memory:")
df_table.to_sql('summary_store', conn, if_exists='replace', index=False)
# pd.read_sql('select * from summary_store', conn)
conn.commit()
conn.close()

# [ENTREGÁVEL] Criação de uma tabela no SQLite em arquivo
conn = sqlite3.connect("../data/db.sqlite")
df_table.to_sql('summary_store', conn, if_exists='replace', index=False)
# pd.read_sql('select * from summary_store', conn)
conn.commit()
conn.close()

# [ENTREGÁVEL] Gravando o dataframe em CSV
df_table.to_csv("../data/summary_store.csv", quoting=csv.QUOTE_NONNUMERIC)

# [ENTREGÁVEL] Gravando o dataframe em JSON
df_table.to_json('../data/summary_store.json', orient="records")